Use fewer buffers and reduce string-copying

UnitedMarsupials-zz opened the pull request:
Also:
 * call `mkstemp(3)` instead of a static name for the action-script
 * Use `syslog(3)` to log errors instead of printing them to `stderr`
 * Use default prompt-strings with `pam_get_authtok` and `pam_get_user`
 * Use more accurate return values in cases of different errors, instead of always replying with `PAM_AUTH_ERR`

UnitedMarsupials-zz committed:
Only build the shared module and name it pam_multipassword.so.

Depend on libcrypto and link with same.

UnitedMarsupials-zz committed:
Use fewer buffers and reduce string-copying a little.

UnitedMarsupials-zz committed:
Further reduce the amount of string-copying

UnitedMarsupials-zz committed:
Avoid filename clashes by using mkstemp instead of static /tmp/action.
Tighten permissions on the file.

UnitedMarsupials-zz committed:
Allow for longer usernames before truncating them.

UnitedMarsupials-zz committed:
Use syslog(3) to log various possible errors. None of these seem like
they should ever be communicated to the user, but, if any should be,
pam_error() should be used for the purpose instead of writing to
stderr directly.

UnitedMarsupials-zz committed:
Tell pam_get_authtok and pam_get_user to use default prompts, if
any is necessary. This saves a modicum of space but also, more
importantly, makes usage of our module harder to detect for any
(potentially hostile) observer.

UnitedMarsupials-zz committed:
Eliminate an empty line, which should not be there.

UnitedMarsupials-zz committed:
Use LOG_AUTH, when calling syslog(3).

UnitedMarsupials-zz committed:
Eliminate more gratuitous casting.

rafket commented:
I disagree with the syslog usage. These errors are very serious and the user should be aware of them upon usage. I agree with the rest of the changes.

UnitedMarsupials-zz committed:
Allow the /usr/share/duress string to be overwritten by compiler define --
/usr/share is read-only on some systems. Use the path /var/db/multipassword
on BSD-systems.

Catch and properly report (with syslog(3)) more potential failures.

UnitedMarsupials commented:
> These errors are very serious and the user should be aware of them upon usage.

Reporting these errors will reveal usage of pam_duress to a potentially hostile party watching from behind the user's shoulder... If they are in a safe environment, they can login as themselves (or root) and find all of the error-messages in the logfile...

At the very least, `PAM_SILENT`-flag should be obeyed -- it controls, whether or not a PAM-module is allowed to produce output (other than through `syslog(3)`) at all. If the flag is not set, then `pam_error(3)` (and/or `pam_info(3)`) should be used to communicate information to the user, not writing directly to `stderr` -- for all we know, the user may be logging in through GUI (such as xdm)...

But best, in my opinion, is to just be silent. :-)

> Also, the casting is necessary, as those variables are of different type.

Some of the casting was not quite right, because it dropped the `const`-qualifier. Instead of fixing it all, I simply removed it altogether -- I normally compile with `-Wno-pointer-sign`, which is a FreeBSD default. I'll add the warning explicitly and fix the calls...

> There is no need to allow larger usernames and then truncate them.

They are only truncated, if they are too long even for what is allowed. In cases of crashes, an admin may have easier time sorting through abandoned `/tmp/action.*.foo`, may he not?

UnitedMarsupials-zz committed:
Cast characters to bytes as appropriate to cleanly compile with -Wpointer-sign.

Lqp1 commented:
Any chance to see this PR merged - at least partially into master ?
